consensus/ibft/ibft.go

```diff
@@ -847,6 +845,14 @@ func (i *Ibft) runAcceptState() { // start new round
 		} else {
 			// since it's a new block, we have to verify it first
 			if err := i.verifyHeaderImpl(snap, parent, block.Header); err != nil {
+				i.logger.Error("block header verification failed", "err", err)
+				i.handleStateErr(errBlockVerificationFailed)
+
+				continue
+			}
+
+			// Verify other block params
+			if err := i.blockchain.VerifyProposedBlock(block); err != nil {
 				i.logger.Error("block verification failed", "err", err)
 				i.handleStateErr(errBlockVerificationFailed)
 
```

Comment:
Can you give me the context why you need to verify block during AcceptState? Is this the improvement that is different from original IBFT flow, right?

Reply:
This is actually the original IBFT spec flow.

The flow we have currently on the `develop` branch, is that you **don't actually verify a block** before committing to it - you're essentially doing consensus on top of a header filed (check out `verifyHeaderImpl` that was the only check in `AcceptState`).
The logic was that if the block is invalid, you'll error out to round change state, which is insane if you consider that you just went through a consensus round for an unverified block.

Calling the block verification here makes sure that the proposed block is actually valid, and that every part of the consensus process is based on a **verified** block.

Commit state is just writing the block down, no need to verify it again since you're sure it's valid.

Reply:
@lazartravica I'd like you to confirm it. I may overlook something

Reply:
We have a Gitbook doc that details and links why this verification needs to be done in `AcceptState` - it's not really a question whether it should be there, but a correction of bad module design.